File: aspnet-core-minimal/src/UserManagement.Api/Repositories/UserRepository.cs
using System.Collections.Concurrent;
using UserManagement.Api.Dtos;

namespace UserManagement.Api.Repositories;

public class UserRepository : IUserRepository
{
    private readonly ConcurrentDictionary<int, User> userDatabase = new (new Dictionary<int, User>
        {
            { 1, new User(1, "Jane", "First user", "jane@test.org") },
            { 2, new User(2, "John", "Second user", "john@test.org") }
        });
    
    private int actualId = 2;
    
    public async Task<IEnumerable<User>> GetAllUsersAsync()
    {
        return await Task.FromResult(userDatabase.Values.ToList()).ConfigureAwait(false);
    }

    public async Task<User?> GetUserByIdAsync(int id)
    {
        if (!userDatabase.TryGetValue(id, out var user))
        {
            return null;
        }
        
        return await Task.FromResult(user).ConfigureAwait(false);
    }

    public async Task<User> CreateUserAsync(CreateUser user)
    {
        var nextId = Interlocked.Increment(ref actualId);
        
        var newUser = new User(nextId, user.Name, user.Description, user.Email);
        
        if (!userDatabase.TryAdd(nextId, newUser))
        {
            throw new InvalidOperationException("Failed to add a new user. The ID might already exist.");
        }

        return await Task.FromResult(newUser).ConfigureAwait(false);
    }

    public Task UpdateUserAsync(User user)
    {
        if (!userDatabase.TryUpdate(user.Id, user, userDatabase[user.Id]))
        {
            throw new InvalidOperationException("Failed to update the user.");
        }
        
        return Task.CompletedTask;
    }

    public Task DeleteUserAsync(int id)
    {
        if (!userDatabase.TryRemove(id, out var user))
        {
            if (user == null)
            {
                throw new InvalidOperationException("User not found.");    
            }
            
            throw new InvalidOperationException("Failed to delete the user.");
        }
        
        return Task.CompletedTask;
    }
}

File: spring-boot/UserManagement/src/main/java/de/biegota/usermanagement/repository/UserRepository.java
package de.biegota.usermanagement.repository;

import de.biegota.usermanagement.model.dto.CreateUser;
import de.biegota.usermanagement.model.dto.User;
import org.springframework.stereotype.Repository;

import java.util.concurrent.ConcurrentHashMap;

@Repository
public class UserRepository implements IUserRepository {

    private final ConcurrentHashMap<Integer, User> userDatabase = new ConcurrentHashMap<>() {{
        put(1, new User(1, "Jane", "First user",  "jane@test.org"));
        put(2, new User(2, "John", "Second user",  "john@test.org"));
    }};

    private int actualId = 2;

    @Override
    public Iterable<User> getAllUsers() {
        return userDatabase.values();
    }

    @Override
    public User getUserById(int id) {
        var user = userDatabase.get(id);
        if (user == null) {
            throw new RuntimeException("User not found");
        }
        return userDatabase.get(id);
    }

    @Override
    public User createUser(CreateUser user) {
        var nextId = actualId++;

        var newUser = new User(nextId, user.Name(), user.Description(), user.Email());

        if (userDatabase.putIfAbsent(nextId, newUser) == null)
        {
            throw new RuntimeException("Failed to add a new user. The ID might already exist.");
        }

        return newUser;
    }

    @Override
    public void updateUser(User user) {
        if (!userDatabase.containsKey(user.Id())){
            throw new RuntimeException("User not found");
        }

        userDatabase.replace(user.Id(), user);
    }

    @Override
    public void deleteUser(int id) {
        if (!userDatabase.containsKey(id)){
            throw new RuntimeException("User not found");
        }

        userDatabase.remove(id);
    }
}
